### PR TITLE
Use pytave for copying variables to python

### PR DESCRIPTION
**Pull request by genuinelucifer**
python `compile` function only works on Python 2.
Theoritically, we can use this function for all the type conversion. But here it is used only for `@sym` objects. Rest of the objects depend on `pycall` to convert them.

### CODE REVIEW

**Comment by genuinelucifer** [comment]
I'm not exactly sure that this won't break anything but I couldnt actually find any more use of `python_header` after copying vars stuff was removed.
And since you yourself commented that calling python_header everytime was sort of waste, I though it would be good to import everything while setting the ipc. Is it ok? Or should I move the library importing stuff into `python_ipc_pytave` back?

File: inst/sympref.m [hunk 1]
```diff
@@ -306,6 +306,9 @@
             msg = 'Choosing the default [autodetect] communication mechanism';
           case 'pytave'
             msg = 'Forcing the PyTave communication mechanism';
+            pyexec(strjoin({'import sympy as sp',
+                            'from sympy import __version__ as spver', 
+                            'from sympy import *'}, newl));
```
Comment (cbm755) [inline inst/sympref.m]: this is not the right place, the first time `python_ipc_pytave` is called, use a `persistent` variable to keep track of its the first time.

Comment (cbm755) [inline inst/sympref.m]: anyway, I can't test becuase `newl` is not defined.  Its kinda rude to ask for review when there are trivial problems.

Comment (cbm755) [inline inst/sympref.m]: oh sorry, you didn't ask for review :)  ignore the comment about rudeness.

**Comment by genuinelucifer** [comment]
Currently the code is more mature. Although since everything is stored as numpy arrays, we face dificulties in many things. Like:
`python_cmd('return 2*_ins[0]', [[2 2], [1 1]])` would work.
But, `python_cmd('return 2*_ins[0]', [[2 2]; [1 1]])` won't give required answer...
.
More changes would probably be beneficial on pytave side rather than on symbolic side.
I do some changes and probably open a PR onto `mtmiller/pytave`.

**Comment by cbm755** [comment]
Nice to see this coming along.  Few comments:

1.  You need something like:
````
  if (strcmp(what, 'reset'))
    have_headers = [];
````

2.  There are lots more imports and workaround that need imported, see `python_header.py`, probably everything between:
````
    from sympy.logic.boolalg import Boolean, BooleanFunction
 ---- 8< ----
    from sympy.functions.special.hyper import TupleArg
````

**Comment by genuinelucifer** [comment]
`[a,b] = python_cmd('return _ins', 1, 2)` works BUT both of  
`a = python_cmd('return _ins', [1, 2])`  and  `a = python_cmd('return _ins', {1, 2})` give only 2 as result and show warning that number of results do not match.
But, weirdly I get error that second member of output list is not defined if I try:  
`[a,b] = python_cmd('return _ins', [1, 2])`  or  `[a,b] = python_cmd('return _ins', {1, 2})`
.
I have tried many many things but I feel like I am causing even more errors than resolving them...
.
Any suggestions?

**Comment by cbm755** [comment]
I'll take a look tonight.  Hang in there, we'll get it!

**Comment by cbm755** [comment]
This doesn't work for me:
````
>> x = sym(1)
OctSymPy v2.4.1-dev: this is free software without warranty, see source.
Using experimental PyTave communications with SymPy.
  File "<string>", line 35
    _temp = 
           ^
SyntaxError: invalid syntax
error: pycall: 'module' object has no attribute 'pyclear'
error: called from
    python_ipc_pytave at line 110 column 3
    python_ipc_driver at line 59 column 13
    python_cmd at line 166 column 9
    sym at line 365 column 5
    sym at line 236 column 7
>> 
````

File: inst/private/python_ipc_pytave.m [hunk 2]
```diff
@@ -59,17 +61,56 @@
...
+                    '            n[k] = a[k]',
+                    '    return n',
+                    'def Version(v):',
+                    '    # short but not quite right: https://github.com/cbm755/octsympy/pull/320',
+                    '    return LooseVersion(v.replace(".dev", ""))',
+                    '_ins = []',
+                    'def pyclear():',
+                    '    global _ins',
+                    '    _ins = []',
+                    '_temp = '
```
Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: wtf is this committed?

Comment (genuinelucifer) [inline inst/private/python_ipc_pytave.m]: Oops. That was supposed to be
_temp=None
But `syms x` worked for me anyway..

**Comment by cbm755** [comment]
Here's a command I used for testing:
````
>> [a,b] = python_cmd('print(_ins); return [1,2]', "one", x, y, "two", {"hello", 30}, 40)
['one', x, y, 'two', ['hello', 30.0], 40.0]
a =  1
b =  2
````
(note we can print in python for debugging)

**Comment by cbm755** [comment]
Another useful thing for testing:
````
>> pycall('pystoretemp', 42)
>> pyexec('print _temp')
[[ 42.]]
````

**Comment by cbm755** [comment]
Does this have the #456 fix?  I looks like it does but it doesn't work:

````
>> L = python_cmd('return (1,2),')
warning: number of outputs don't match, was this intentional?
warning: called from
    python_cmd at line 188 column 5
L =  1
>> 
````

**Comment by genuinelucifer** [comment]
It should have that fix. I cherry picked from here only. Probably I messed
something else too. I'll try to fix it.

File: inst/private/python_ipc_pytave.m [hunk 3]
```diff
@@ -59,17 +61,50 @@
...
+                    'from sympy.matrices.expressions.matexpr import MatrixElement',
+                    '# for hypergeometric',
+                    'from sympy.functions.special.hyper import TupleArg',
+                    'from sympy.utilities.iterables import uniq',
+                    'import copy',
+                    'import binascii',
+                    'import struct',
+                    'import codecs',
+                    'import xml.etree.ElementTree as ET',
+                    'from distutils.version import LooseVersion',
```
Comment (genuinelucifer) [inline inst/private/python_ipc_pytave.m]: are all these imports required? I merely copy pasted all of them because some tests showed error regarding a few of the imports..

Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: You should be able to drop `xml`, `binascii`.

Probably/maybe `codecs`, `copy`, `struct` too.

The SymPy stuff is probably all needed: this is because these things appear in `srepr` for some SymPy object, so we need them to copy objects *back* to SymPy.   Worse, they need to be in the main namespace...  Possibly `git blame inst/private/python_header.py` will tell you more about each of them.

**Comment by cbm755** [comment]
I've tried this again: it seems much worse than the `pytave` branch right now, lots of things are failing on *return* from SymPy.  But this branch is about copying *into* Python...?   What is happening?

`findsymbols` failing for example...

**Comment by genuinelucifer** [comment]
> lots of things are failing on return from SymPy. But this branch is about copying into Python...? What is happening?

Hmm... Sorry I forgot to deal with this branch in the past week. I'll try to see today what exactly is going wrong. And yes, now that you mention it, since it is just copying into python the return statements should not cause octave to crash (I am getting octave crash on many tests).

File: inst/private/python_ipc_pytave.m (hunk 3 above)
Comment (genuinelucifer) [inline inst/private/python_ipc_pytave.m]: Thanks. I did `git blame` but it only shows which changes in this file were made when/by whome.
But, some imports, although not required by `python_header.py`,are still required by some tests...
Since the imports are not that many, I'll try to remove them one by one and try.

Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: Just FYI, let me walk you through it:

`git blame python_header.py`

````
---- 8< ----
7b6c35b5 (Colin B. Macdonald 2014-11-16 15:23:32 +0000  35)     from sympy.functions.elementary.piecewise import ExprCondPair
51a80384 (Colin B. Macdonald 2016-01-07 00:26:37 -0800  36)     from sympy.integrals.risch import NonElementaryIntegral
9474bd5a (Colin B. Macdonald 2016-01-11 00:05:02 -0800  37)     from sympy.matrices.expressions.matexpr import MatrixElement
400811d8 (Colin B. Macdonald 2016-05-09 13:04:12 -0700  38)     # for hypergeometric
400811d8 (Colin B. Macdonald 2016-05-09 13:04:12 -0700  39)     from sympy.functions.special.hyper import TupleArg
af6e79e7 (latot              2016-01-17 02:01:21 -0300  40)     from sympy.utilities.iterables import uniq
---- 8< ----
````

So lets look at this one: `from sympy.integrals.risch import NonElementaryIntegral`:

`git show 51a80384`

````
commit 51a80384d5caea0db211e452d39f0b4f6b3778cc
Author: Colin B. Macdonald <cbm@m.fsf.org>
Date:   Thu Jan 7 00:26:37 2016 -0800

    Fix passing NonElementaryIntegrals back to Python
    
    Fixes https://savannah.gnu.org/bugs/index.php?46831.
    
    Yet another case where `eval(srepr(...))` is not quiet the
    identity map.
````

(this doesn't always work b/c the most recent change to a line might not explain why that line is there...)

Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: hmmm "not **quite** the identity map": someone is not proofreading his commits :-)

**Comment by genuinelucifer** [comment]
@cbm755 This feels like a lucky day!  ^_^
Even this branch is looking good now. I got no crashes and only 25 tests failed! :)
Looks like we will be merging this branch soon.

**Comment by genuinelucifer** [comment]
This is really going great! :D
Now, we have ONLY 9 failing tests!!
.
3 tests fail in `@sym/solve.m` which will be fixed when #486 is done.
6 tests fail in `python_cmd.m` which are due to escaping characters in strings..
.
I can fix the tests in `@sym/solve.m` but I have no clue how to fix tests in `python_cmd.m`...
Ex (One of the tests that fails):
```
>> x = 'a string\nbroke off\nmy guitar\n';
>> x2 = sprintf(x);
>> y = python_cmd ('return _ins', x);
a string\nbroke off\nmy guitar\n
>> x3 = strrep(x2, sprintf('\n'), sprintf('\r\n'));  % windows
>> assert (strcmp(y, x2) || strcmp(y, x3))
error: assert (strcmp (y, x2) || strcmp (y, x3)) failed
error: called from
    assert at line 92 column 11
>> x2
x2 = a string
broke off
my guitar

>> x3
x3 = a string
broke off
my guitar

>> y
y = a string\nbroke off\nmy guitar\n
```

Any pointers?

**Comment by cbm755** [comment]
I'll look into `python_cmd`: there is at least one weird test in there...

**Comment by cbm755** [comment]
I filed https://bitbucket.org/mtmiller/pytave/issues/24/pyexec-inconsistent-error-handling about some of the error handling stuff.

re: the string failures: feels like we are doing the right thing: those tests look more like enforcing the limitations of the other IPC mechanisms.  I'll deal with that.

**Comment by genuinelucifer** [comment]
Thanks. :)
I'll probably sleep now. Will push the fix for dictionary tomorrow.

**Comment by cbm755** [comment]
pull from `pytave` to fix the string stuff

**Comment by genuinelucifer** [comment]
Now, only 2 tests fail in python_cmd that too they are expected to fail. But they expect an error message while they get a python exception.
Probably lowest priority for now. #457 is the cause.

**Comment by cbm755** [comment]
For the "very specialized test", I suggest making it `%!xtest error <AttributeError>` and adding a comment to the pytave bug.

For the "export type" one, that probably should work with `@pyobject`.  Could just xtest for now with a note that it will succeed on pytave.

**Comment by genuinelucifer** [comment]
We now have only 2 tests failing. Changing them to xtest will fail the other IPCs that's why I didn't change them.... Should I?

**Comment by cbm755** [comment]
either passing or failing an `xtest` is ok (this differs from doctest XFAIL which must fail).

**Comment by cbm755** [comment]
This PR is getting a bit unwieldy.  Can you submit the xtest change we just talked about to `pytave` branch as a new PR?  I think that is unrelated to this PR.  Same for any other changes related to python-to-octave  (dict stuff?).

**Comment by genuinelucifer** [comment]
I removed the 2 commits from this branch. And submitted 2 PRs...
.
Also, Should I merge commits on this PR?
Because I guess we have covered all the cases (that are currently being used)...

File: inst/private/store_vars_in_python.m [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      pyexec([varname '.append(' char(x) ')'])
+    elseif (iscell (x))
+      tempname = [varname num2str(counter)];
+      counter = counter + 1;
+      pyexec ([tempname ' = []'])
+      store_vars_in_python (tempname, x)
+      pyexec([varname '.append(' tempname ')'])
+    elseif (isscalar (x) && isnumeric (x))
+      % workaround upstream PyTave bug: https://fixme.url.here
+      % FIXME: add bug number, about making arrays [[42]] when passing 42
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: please do these fixme's

File: inst/private/store_vars_in_python.m [hunk 5]
```diff
@@ -0,0 +1,31 @@
+function store_vars_in_python (varname, L)
+%varname
+%L
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: remove

File: inst/private/store_vars_in_python.m [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  persistent counter = round(100000*rand())
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: I noted this before somewhere but its still here: why we have this rand number?  Can just set to 0?

File: inst/private/store_vars_in_python.m [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      counter = counter + 1;
+      pyexec ([tempname ' = []'])
+      store_vars_in_python (tempname, x)
+      pyexec([varname '.append(' tempname ')'])
+    elseif (isscalar (x) && isnumeric (x))
+      % workaround upstream PyTave bug: https://fixme.url.here
+      % FIXME: add bug number, about making arrays [[42]] when passing 42
+      pycall ('pystoretemp', x)
+      if isinteger(x)
+        pyexec ([varname '.append(int(_temp[0,0]))'])
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: Can you add something like:
````
% FIXME: is this working around a PyTave bug?  without it type is e.g., numpy.uint8
````

File: inst/private/store_vars_in_python.m [hunk 8]
```diff
@@ -0,0 +1,31 @@
+function store_vars_in_python (varname, L)
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: Can you put a GNU copyright header on this?

**Comment by cbm755** [comment]
> Also, Should I merge commits on this PR?

You mean squash I guess.  Do you think it needs it?   It looks ok to me.  Are there any binary files or terribly embarrasing code in there? ;-)  If not, its fine as is for me.

File: inst/private/store_vars_in_python.m (hunk 8 above)
Comment (genuinelucifer) [inline inst/private/store_vars_in_python.m]: Done

**Comment by genuinelucifer** [comment]
> You mean squash I guess.

Oops, yes... Seems ok to me too. won't squash..

File: inst/private/store_vars_in_python.m [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      tempname = [varname num2str(counter)];
+      counter = counter + 1;
+      pyexec ([tempname ' = []'])
+      store_vars_in_python (tempname, x)
+      pyexec([varname '.append(' tempname ')'])
+    elseif (isscalar (x) && isnumeric (x))
+      % workaround upstream PyTave bug: https://bitbucket.org/mtmiller/pytave/issues/14
+      pycall ('pystoretemp', x)
+      if isinteger(x)
+	    % workaround as PyTave apparently stores everything as Float
```
Comment (cbm755) [inline inst/private/store_vars_in_python.m]: no, its more complicated then that.  It goes to `numpy.uint8` or someother `numpy` type.  your `int` here brings it to native python int type.  please leave an explicit "FIXME" until we decide if that is pytave problem or not.

File: inst/private/python_ipc_pytave.m (hunk 3 above)
Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: > You should be able to drop xml, binascii.

*ping*

File: inst/private/python_ipc_pytave.m [hunk 10]
```diff
@@ -59,17 +61,50 @@
...
+                    'def dictdiff(a, b):',
+                    '    """ keys from a that are not in b, used by evalpy() """',
+                    '    n = dict()',
+                    '    for k in a:',
+                    '        if not k in b:',
+                    '            n[k] = a[k]',
+                    '    return n',
+                    'def Version(v):',
+                    '    # short but not quite right: https://github.com/cbm755/octsympy/pull/320',
+                    '    return LooseVersion(v.replace(".dev", ""))',
```
Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: can we remove the commas from the end of these lines?  some have some don't, I guess I prefer none.

File: inst/private/python_ipc_pytave.m [hunk 11]
```diff
@@ -59,17 +61,50 @@
...
-  loc = python_copy_vars_to('_ins', false, varargin{:});
-  s = strjoin(loc, newl);
-  pyexec(s)
+  pyexec('_ins = []');
+  store_vars_in_python('_ins', varargin);
 
   s = strjoin(cmd, newl);
-  pyexec(s)
-
+  pyexec(s);
```
Comment (cbm755) [inline inst/private/python_ipc_pytave.m]: unnecessary change

File: inst/private/python_ipc_pytave.m (hunk 3 above)
Comment (genuinelucifer) [inline inst/private/python_ipc_pytave.m]: sorry, never got around to test it. I'll do it now.

**Comment by genuinelucifer** [comment]
@cbm755 Done. I hope didn't miss anything..

**Comment by cbm755** [comment]
Cool, I'll plan to merge this when tests pass (nice to be able to say that again!)

(If the tests fail, try pulling from `pytave`.)

**Comment by genuinelucifer** [comment]
Well, 3 tests will fail... But, after merging to `pytave` they won't fail. Due to the dicts fix...
.
Should I merge pytave here?

**Comment by cbm755** [comment]
Yeah merge here, it keeps segfaulting.  (I wasn't sure if travis runs this branch or the result of this branch, merged to the target).

**Comment by genuinelucifer** [comment]
I saw the segfaults now. On my pc (without merge), the tests were running fine (with 3 failing tests)..
Donno what happened on travis...

**Comment by genuinelucifer** [comment]
@cbm755 Any pointers here? I have no clue what is going wrong. None of the changes seem to drastic to cause a segfault!

**Comment by genuinelucifer** [comment]
@cbm755 Any pointers? I am not able to figure out anything concrete which might be causing the segfaults...

**Comment by cbm755** [comment]
Possibly this branch depends on pyobect somehow?

**Comment by cbm755** [comment]
What's the status of this?  Pytave does a lot more now, it would be good to update this if necessary.  I'd like to merge this to my pytave branch (and hopefully soon to master).

(I should think about #506 but wanted to check if this is up-to-date first)

**Comment by genuinelucifer** [comment]
Well, for me most of the case pass. I don't think we'll need changes here. Probably we can drop a few things.

And as you said travis only has 4.0.x so we can't test there. I'll try to update this (if needed) by evening (~5-6hrs)..

**Comment by cbm755** [comment]
Travis support is not urgent.  Please test this branch against pytave master and let me know if you think we can start to merge things.

**Comment by genuinelucifer** [comment]
I have done some more testing and it seems that we can safely merge this to pytave after #507 is resolved upstream...

**Comment by cbm755** [comment]
Can you update `check_and_convert` and the to use the newest Pytave features?  Also, add a copy copyright header?

Re: dict: maybe for now, we should [convert them to structs](https://bitbucket.org/mtmiller/pytave/issues/62).

**Comment by genuinelucifer** [comment]
@cbm755 I will send a new PR for the changes in `check_and_convert` as PR was only for `store_vars_in_python`...

Please review this and merge if it's ok.. I'll send the other PR with the dict change in a couple of hours.

**Comment by cbm755** [comment]
Ok, I'll merge this to the `pytave` branch.  I think the import can be re-written with newer pytave features but we can do it in a new PR.
